framework/src/org/checkerframework/framework/util/MultiGraphQualifierHierarchy.java

```diff
@@ -236,10 +236,28 @@ protected void finish(QualifierHierarchy qualHierarchy,
     public String toString() {
         // TODO: it would be easier to debug if the graph and map were sorted by the key.
         // Simply creating a TreeMap here doesn't work, because AnnotationMirrors are not comparable.
-        return "Supertypes Graph: " + supertypesGraph.toString() +
-                "\nSupertypes Map: " + String.valueOf(supertypesMap) +
-                "\nTops: " + tops +
+
+        String output = "Supertypes Graph: ";
+        for (AnnotationMirror key : supertypesGraph.keySet()) {
+            output += "\n\t" + key.toString() + "=" + supertypesGraph.get(key);
+        }
+
+        output += "\nSupertypes Map: ";
+        for (AnnotationMirror mapKey : supertypesMap.keySet()) {
+            output += "\n\t" + mapKey.toString() + "=[";
+            if (!supertypesMap.get(mapKey).isEmpty()) {
+                for (AnnotationMirror mapValKey : supertypesMap.get(mapKey)) {
+                    output += mapValKey.toString() + ", ";
+                }
+                output = output.substring(0, output.length() - 2);
```

Comment:
I find this a rather ugly way to concatenate the list elements with a separator. Can you write the if/for parts to be easier to read?